server/src/test/java/org/elasticsearch/env/NodeMetaDataTests.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.env;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.EqualsHashCodeTestUtils;
+import org.elasticsearch.test.VersionUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.endsWith;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.startsWith;
+
+public class NodeMetaDataTests extends ESTestCase {
+    private Version randomVersion() {
+        return rarely() ? Version.fromId(randomInt()) : VersionUtils.randomVersion(random());
+    }
+
+    public void testEqualsHashcodeSerialization() {
+        final Path tempDir = createTempDir();
+        EqualsHashCodeTestUtils.checkEqualsAndHashCode(new NodeMetaData(randomAlphaOfLength(10), randomVersion()),
+            nodeMetaData -> {
+                final long generation = NodeMetaData.FORMAT.writeAndCleanup(nodeMetaData, tempDir);
+                final Tuple<NodeMetaData, Long> nodeMetaDataLongTuple
+                    = NodeMetaData.FORMAT.loadLatestStateWithGeneration(logger, xContentRegistry(), tempDir);
+                assertThat(nodeMetaDataLongTuple.v2(), equalTo(generation));
+                return nodeMetaDataLongTuple.v1();
+            }, nodeMetaData -> {
+                if (randomBoolean()) {
+                    return new NodeMetaData(randomAlphaOfLength(21 - nodeMetaData.nodeId().length()), nodeMetaData.nodeVersion());
+                } else {
+                    return new NodeMetaData(nodeMetaData.nodeId(), randomValueOtherThan(nodeMetaData.nodeVersion(), this::randomVersion));
+                }
+            });
+    }
+
+    public void testReadsFormatWithoutVersion() throws IOException {
+        // the behaviour tested here is only for compatibility with versions 7 and earlier
+        // so it (and testReadsFormatWithoutVersion.binary) can be removed once this compatibility is no longer required
+        assertTrue(Version.CURRENT.minimumIndexCompatibilityVersion().onOrBefore(Version.V_7_0_0));
+
+        final Path tempDir = createTempDir();
+        final Path stateDir = Files.createDirectory(tempDir.resolve(MetaDataStateFormat.STATE_DIR_NAME));
+        final InputStream resource = this.getClass().getResourceAsStream("testReadsFormatWithoutVersion.binary");
```

Comment:
I would rather avoid adding a binary file a resource. Can we have instead a method for creating NodeMetadData without version?

Reply:
I would rather do it this way, because the important behaviour is that we can correctly read versions of this file written by arbitrarily ancient versions of Elasticsearch of which the given resource is an example. If we create the metadata file in this test then we might change the format in a way that round-trips correctly but which doesn't reflect the behaviour of ancient versions of Elasticsearch any more.

Reply:
Isn't this tested by our full / rolling cluster upgrade tests?

Reply:
Today it is, but in future we won't be able to produce version-free metadata files any more and yet we'll want to have a test saying that we refuse to load them in place of the one here. However the comment I left is misleading and I adjusted it in 9cb083f6605881b2141e844f97a414485f3b0a14.